public/js/widget_config.js

```diff
@@ -0,0 +1,382 @@
+
+  // TODO: Figure out how to replace the string 'widget' with a constant
+  const widgetContainer = jQuery(
+    `<div class="widget ${widgetTypeUpper}" id="${objWidget.label}"></div>`
+  )
+  const widgetHeader = jQuery(
+    '<div class="widget-header">' + objWidget.label + '</div>'
+  )
+  const widgetKebobMenu = jQuery(
+    '<img class="widget-kebobMenu" src="img/kebobMenu.png"/>'
+  )
+  const widgetContent = jQuery(
+    '<div class="widget-content"></div>'
+  )
+
+  widgetKebobMenu.appendTo(widgetContainer)
+  widgetHeader.appendTo(widgetContainer)
+  widgetContent.appendTo(widgetContainer)
+
+  if (Object.hasOwn(objWidget, 'keys')) {
+    keyControl.addKeysForWidget(objWidget) // eslint-disable-line no-undef
+  }
+
+  /*
+   * Store the widget configuration object (objWidget) under the WIDGET_NAMESPACE
+   * in an arbetrary jQuery data unit attatched to the widgetContainer
+   */
+  jQuery(widgetContainer).data(RQ_PARAMS.WIDGET_NAMESPACE, objWidget)
+
+  console.debug(`${JSON.stringify(objWidget)}`)
+
+  jQuery(widgetContainer)[widgetTypeUpper](
+    { ...objWidget, socket: socket }
+  ).appendTo(
+    '#widgets'
+  ).draggable({
+    handle: '.widget-header',
+    snap: true,
+    start: function (event, ui) {
+      const widgetId = event.currentTarget.id
+      console.debug(
+        `drag started on ${widgetId}` +
+        ` at position ${JSON.stringify(jQuery('#' + widgetId).position())}` +
+        ` and offset ${JSON.stringify(jQuery('#' + widgetId).offset())}`
+      )
+    },
+    stop: function (event, ui) {
+      const widgetId = event.target.id
+      const widgetData = jQuery('#' + widgetId).getWidgetConfiguration()
+
+      if (!widgetData) {
+        /*
+         * The widgetData no longer exists, likely because the widget was deleted from the UI. That condition
+         * is acceptable here.
+         */
+        return
+      }
+
+      const widgetPosition = widgetData.position
+
+      console.debug(
+        `drag stopped on ${widgetId}` +
+        ` at offset ${JSON.stringify(jQuery('#' + widgetId).offset())}` +
+        ` and position ${JSON.stringify(jQuery('#' + widgetId).position())}` +
+        ` original ${JSON.stringify(widgetPosition)}`
+      )
+
+      jQuery('#' + widgetId).getWidgetConfiguration().position = updateWidgetPosition(
+        widgetPosition,
+        jQuery('#' + widgetId).position()
+      )
+      console.debug(
+        ` updated position to: ${JSON.stringify(jQuery('#' + widgetId).getWidgetConfiguration().position)}`
+      )
+    }
+  }).hover(function (event) {
+    jQuery(event.currentTarget).find('.widget-kebobMenu')[0].style.display = 'block'
+  },
+  function (event) {
+    jQuery(event.currentTarget).find('.widget-kebobMenu')[0].style.display = 'none'
+  })
+
+  /*
+  * Two alternative ways to do the same thing.
+  * We should see what students like best. Double clicking the widget header, or clicking on a kebob menu.
+  * I (Mark) personaly like the double-click option
```

Comment:
Can we find any UI or UX Best Practices covering "how do I change this thing on my screen"?